### PR TITLE
Enh: Add notifications

### PR DESCRIPTION
**Pull request by felixhahnweilheim**
- Notify users when their account has been verified.
- Notify space owners when their space has been verified.

### CODE REVIEW

**Comment by felixhahnweilheim** [comment]
It works:
Screenshot 2023-04-29 at 23-33-18 Protokollierung - Administration - Felix Hahn Development

**To be done**:
- [ ] fix the positioning with the space icon
- [ ] seperate functions in ConfigureForm
- [ ] ev. use afterSave() instead of save()

File: models/ConfigureForm.php [hunk 1]
```diff
@@ -154,6 +160,22 @@ public function saveSettings()
         $settings->setSerialized('verifyUser', (array)$this->verifyUser);
         $settings->setSerialized('verifySpace', (array)$this->verifySpace);
 
+        //Send notification to new verified users
+        $newUsersGuid = array_diff((array)$this->verifyUser, $oldVerifyUsers);
+        $newUsers = array();
```
Comment (ArchBlood) [inline models/ConfigureForm.php:165]: Prefer short arrays `[]`

Comment (felixhahnweilheim) [reply models/ConfigureForm.php]: done

**Comment by felixhahnweilheim** [comment]
Screenshot 2023-05-01 at 23-17-24 Übersicht - Felix Hahn Development

Using from() and about() the layout is now correct. The profile image of the admin is shown and for spaces the space image (small).

**Comment by ArchBlood** [comment]
With this implementation we probably should also implement an integrity check as well.

**Comment by felixhahnweilheim** [comment]
For which part exactly. I've only added one simple setting.
And the notifications are handled by HumHub core. While developping I saw a few warnings like 'invalid notification was deleted'.

Something else: It's recommended to send the notifications from afterSave(). But somehow I didn't get that to work. I don’t get any error, it just doesn't do it.

And maybe codeception tests would be good...

**Comment by ArchBlood** [comment]
> For which part exactly? I've only added one simple setting.
> And the notifications are handled by HumHub core. While developping I saw a few warnings like 'invalid notification was deleted'.
> 
> Something else: It's recommended to send the notifications from afterSave(). But somehow I didn't get that to work. I don’t get any error, it just doesn't do it.
> 
> And maybe codeception tests would be good...

The current implementation is using a similar concept to the update message from my viewing of the code, I've been trying to get the notifications to fire in production without success. I'll look into this more.

**Comment by felixhahnweilheim** [comment]
Just to mention, the notification won't be send to yourself.

**Comment by ArchBlood** [comment]
> Just to mention, the notification won't be send to yourself.

I've also tested on a demo user which didn't receive the notification.

**Comment by felixhahnweilheim** [comment]
I think I have found the reason. I wanted to keep the feature simple and did not add our own notification category. So the default targets are used.
But the default excludes MobileTarget, see here:
https://github.com/humhub/humhub/blob/master/protected/humhub/modules/notification/components/NotificationCategory.php#L67

Should I add our own category? Or maybe use an existing one?

**Comment by ArchBlood** [comment]
> I think I have found the reason. I wanted to keep the feature simple and did not add our own notification category. So the default targets are used. But the default excludes MobileTarget, see here: https://github.com/humhub/humhub/blob/master/protected/humhub/modules/notification/components/NotificationCategory.php#L67
> 
> Should I add our own category? Or maybe use an existing one?

Implementing our own category would be for the best here.

**Comment by ArchBlood** [comment]
After this is implemented, in the Module.php probably need to also add the following;

```php
    /**
     * @inheritdoc
     */
    public function getNotifications()
    {
        return [
            notifications\UserVerified::class,
            notifications\SpaceVerified::class
        ];
    }
```

**Comment by felixhahnweilheim** [comment]
Ok, I thought it is a bit too much to have an own category. Maybe we should hide it for the users? And just have the default admin setting?

And of course, I should now remove the sendNotification setting...

Adding getNotifications() is not needed, the default works fine.

**Comment by ArchBlood** [comment]
Regarding `afterSave()` maybe this could help you in this?
https://github.com/humhub/reportcontent/blob/master/models/ReportContent.php#L123-L143

**Comment by felixhahnweilheim** [comment]
I have just tested and mobile browser notifications also work. (Currently I can't test with the FCM Push notifications module because it does not work in my instance.)

By default e-mail and web are active, mobile inactive.

I would suggest to hide that setting to users. Normally a user gets such a notification only once and we should not confuse them with an unneeded setting. It should be enough that administrators can decide about the setting for everyone. Do you agree?

`afterSave` seems to work when I set the table name to "setting". Should I do that? Or should we just leave it in the `saveSettings()` method.

**Comment by ArchBlood** [comment]
> I have just tested and mobile browser notifications also work. (Currently I can't test with the FCM Push notifications module because it does not work in my instance.)
> 
> By default e-mail and web are active, mobile inactive.

I've tested with FCM and it seems to work fine.

> I would suggest to hide that setting to users. Normally a user gets such a notification only once and we should not confuse them with an unneeded setting. It should be enough that administrators can decide about the setting for everyone. Do you agree?

That would be for the best, another option would be probably marking the check lists `read-only`? 🤔

> `afterSave` seems to work when I set the table name to "setting". Should I do that? Or should we just leave it in the `saveSettings()` method.

Probably for the best to leave it in `saveSettings()` as it still works as intended, unless we can find a reason it would be better to place it under this table name, my fear adding it to this table would be the [commands](https://docs.humhub.org/docs/admin/console/#settings) for the `setting` table. 🤔

**Comment by felixhahnweilheim** [comment]
I don't see a `read-only` option, so I have just hidden them in the account settings.

Just to be clear, all the module settings are saved in the setting table because we use the setting manager `Yii::$app->getModule('verified')->settings->set()`. But this really works fine, it is saved with the module id. (That's also why module is removed when the module is deactivated).

From my side this PR seems ready to be merged.

**Pull request merged**